bin/ghe-backup

```diff
@@ -159,59 +156,54 @@ bm_init > /dev/null
 ghe-backup-store-version  ||
 echo "Warning: storing backup-utils version remotely failed."
 
-echo "Backing up GitHub settings ..."
-ghe-backup-settings || failures="$failures settings"
-
-echo "Backing up SSH authorized keys ..."
-bm_start "ghe-export-authorized-keys"
-ghe-ssh "$GHE_HOSTNAME" -- 'ghe-export-authorized-keys' > authorized-keys.json ||
-failures="$failures authorized-keys"
-bm_end "ghe-export-authorized-keys"
-
-echo "Backing up SSH host keys ..."
-bm_start "ghe-export-ssh-host-keys"
-ghe-ssh "$GHE_HOSTNAME" -- 'ghe-export-ssh-host-keys' > ssh-host-keys.tar ||
-failures="$failures ssh-host-keys"
-bm_end "ghe-export-ssh-host-keys"
-
-echo "Backing up MySQL database ..."
-ghe-backup-mysql || failures="$failures mysql"
-
-echo "Backing up Redis database ..."
-ghe-backup-redis > redis.rdb || failures="$failures redis"
-
-echo "Backing up audit log ..."
-ghe-backup-audit-log || failures="$failures audit-log"
-
-echo "Backing up hookshot logs ..."
-ghe-backup-es-hookshot || failures="$failures hookshot"
-
-echo "Backing up Git repositories ..."
-ghe-backup-repositories || failures="$failures repositories"
-
-echo "Backing up GitHub Pages ..."
-ghe-backup-pages || failures="$failures pages"
-
-echo "Backing up storage data ..."
-ghe-backup-storage || failures="$failures storage"
+# Execute long running backup jobs in parallel
+parallel_commands="
+ghe-backup-settings
+ghe-ssh \"$GHE_HOSTNAME\" -- 'ghe-export-authorized-keys' > authorized-keys.json
+ghe-ssh \"$GHE_HOSTNAME\" -- 'ghe-export-ssh-host-keys' > ssh-host-keys.tar
+ghe-backup-mysql
+ghe-backup-redis > redis.rdb
+ghe-backup-audit-log;
+ghe-backup-es-hookshot
+ghe-backup-repositories
+ghe-backup-pages
+ghe-backup-storage
+ghe-backup-git-hooks
+"
+
+echo "Backing up GitHub settings ...
+Backing up SSH authorized keys ...
+Backing up SSH host keys ...
+Backing up MySQL database ...
+Backing up Redis database ..
+Backing up audit log ...
+Backing up hookshot logs ...
+Backing up Git repositories ...
+Backing up GitHub Pages ...
+Backing up storage data ...
+Backing up custom Git hooks ..."
```

Comment:
I think now all verbose log are piped into `GHE_VERBOSE_LOG` which is a single file. Once we move this to parallel, I am worried about the verbose log will be all messed up and impossible to read. We probably should start to separate the verbose log file now.

Reply:
Each tasks logs will be outputted together one after another as tasks are completed. Tasks will NOT inline output with one another . Therefore verbose logs should maintain their readability. This is one of the primary reasons why I chose this utility.

Reply:
For good measure I went ahead and added the following phrase  to the beginning of each task.
`ghe_verbose "==== ghe-backup-task ===="`

Reply:
I fixed this by using the following:
```
calling_script_name="$(caller | sed 's:.*/::')"
exec 3> >(sed "s/^/[$(date "+%Y-%m-%d %H:%M:%S")] $calling_script_name: /" >> "$GHE_VERBOSE_LOG")
```
The results look something like this:
```
[2020-04-01 08:50:38] ghe-backup-settings: * Transferring settings data ...
[2020-04-01 08:50:38] ghe-backup-settings: * Transferring license data ...
[2020-04-01 08:50:38] ghe-backup-settings: * Transferring management console password ...
[2020-04-01 08:50:38] ghe-backup-settings: * Transferring CA certificates ...
[2020-04-01 08:50:38] ghe-backup-settings: * Transferring UUID ...
[2020-04-01 08:50:38] ghe-backup-es-rsync: * Performing initial sync of ES indices ...
[2020-04-01 08:50:38] ghe-backup-es-rsync: receiving incremental file list
[2020-04-01 08:50:38] ghe-backup-es-rsync: ./
[2020-04-01 08:50:38] ghe-backup-es-rsync: 
[2020-04-01 08:50:38] ghe-backup-es-rsync: sent 169 bytes  received 8,585 bytes  17,508.00 bytes/sec
[2020-04-01 08:50:38] ghe-backup-es-rsync: total size is 121,190  speedup is 13.84
[2020-04-01 08:50:38] ghe-backup-es-rsync: * Disabling ES index flushing ...
[2020-04-01 08:50:38] ghe-backup-es-rsync: * Performing follow-up sync of ES indices ...
[2020-04-01 08:50:38] ghe-backup-es-rsync: receiving incremental file list
[2020-04-01 08:50:38] ghe-backup-es-rsync: 
[2020-04-01 08:50:38] ghe-backup-es-rsync: sent 162 bytes  received 8,578 bytes  17,480.00 bytes/sec
[2020-04-01 08:50:38] ghe-backup-es-rsync: total size is 121,190  speedup is 13.87
[2020-04-01 08:50:38] ghe-backup-es-rsync: * Enabling ES index flushing ...
[2020-04-01 08:50:38] ghe-backup-pages: 
[2020-04-01 08:50:38] ghe-backup-pages: * Starting backup for host: 10.0.1.134
[2020-04-01 08:50:38] ghe-backup-pages: 
[2020-04-01 08:50:38] ghe-backup-pages: * Transferring pages files ...
[2020-04-01 08:50:38] ghe-backup-pages: receiving incremental file list
[2020-04-01 08:50:38] ghe-backup-pages: ./
[2020-04-01 08:50:38] ghe-backup-pages: 
[2020-04-01 08:50:38] ghe-backup-pages: sent 35 bytes  received 226 bytes  522.00 bytes/sec
[2020-04-01 08:50:38] ghe-backup-pages: total size is 0  speedup is 0.00
[2020-04-01 08:50:38] ghe-backup-git-hooks: git-hooks environment tarballs not found. Skipping ...
[2020-04-01 08:50:38] ghe-backup-git-hooks: receiving incremental file list
[2020-04-01 08:50:38] ghe-backup-git-hooks: ./
[2020-04-01 08:50:38] ghe-backup-git-hooks: 
[2020-04-01 08:50:38] ghe-backup-git-hooks: sent 27 bytes  received 52 bytes  158.00 bytes/sec
[2020-04-01 08:50:38] ghe-backup-git-hooks: total size is 0  speedup is 0.00
```